projects/app.py

```diff
@@ -0,0 +1,323 @@
+from config.flask_config import app
+from models.models import Project, Milestone, ReservedOffset, db
+from flask import abort, request, jsonify
+from datetime import datetime
+from classes.response import ResponseBodyJSON
+from classes.exception import CustomException
+from classes.amqp_payload import Payload
+from sqlalchemy.exc import IntegrityError
+from config.amqp_setup import *
+import json
+
+@app.errorhandler(404)
+def handle_resource_not_found(e):
+    """Handles exception thrown when flask app is aborted using `abort` with HTTP status code 404."""
+    app.logger.exception(e)
+    data = CustomException(str(e)).json()
+    res = ResponseBodyJSON(False, data).json()
+    return jsonify(res), 404
+
+@app.errorhandler(400)
+def handle_bad_request(e):
+    """Handles exception thrown when flask app is aborted using `abort` with HTTP status code 400."""
+    app.logger.exception(e)
+    data = CustomException(str(e)).json()
+    res = ResponseBodyJSON(False, data).json()
+    return jsonify(res), 400
+
+@app.errorhandler(KeyError)
+def handle_key_error(e: KeyError):
+    """Handles exception thrown when KeyError is thrown most probably when a field is missing in request body."""
+    app.logger.exception(e)
+    data = CustomException(f"Fields missing from request body {e.args}").json()
+    res = ResponseBodyJSON(False, data).json()
+    return jsonify(res), 400
+
+@app.errorhandler(IntegrityError)
+def handle_integrity_error(e: IntegrityError):
+    """Handles exception when integrity constraint in database are violated. E.g. not nullable fields are null (due to missing fields in request body)"""
+    # print(f'IntegrityError: {e._message}')
+    app.logger.exception(e._message)
```

Comment:
is there a reason for using SQLAlchemyError's private method `_message` instead of just `str(e)` (invokes `__str__()`)?


Taken from [SQLAlchemy's source](https://github.com/zzzeek/sqlalchemy/blob/main/lib/sqlalchemy/exc.py#L74):
```python
class SQLAlchemyError(HasDescriptionCode, Exception):
    """Generic error class."""

    def _message(self) -> str:
        # rules:
        #
        # 1. single arg string will usually be a unicode
        # object, but since __str__() must return unicode, check for
        # bytestring just in case
        #
        # 2. for multiple self.args, this is not a case in current
        # SQLAlchemy though this is happening in at least one known external
        # library, call str() which does a repr().
        #
        text: str

        if len(self.args) == 1:
            arg_text = self.args[0]

            if isinstance(arg_text, bytes):
                text = compat.decode_backslashreplace(arg_text, "utf-8")
            # This is for when the argument is not a string of any sort.
            # Otherwise, converting this exception to string would fail for
            # non-string arguments.
            else:
                text = str(arg_text)

            return text
        else:
            # this is not a normal case within SQLAlchemy but is here for
            # compatibility with Exception.args - the str() comes out as
            # a repr() of the tuple
            return str(self.args)

    def _sql_message(self) -> str:
        message = self._message()

        if self.code:
            message = "%s %s" % (message, self._code_str())

        return message

    def __str__(self) -> str:
        return self._sql_message()
```